subprojects/execution/src/main/java/org/gradle/internal/execution/history/impl/DefaultExecutionHistoryStore.java

```diff
@@ -39,6 +40,7 @@
 public class DefaultExecutionHistoryStore implements ExecutionHistoryStore {
 
     private final PersistentIndexedCache<String, PreviousExecutionState> store;
+    private final boolean isReadOnly = Objects.equals(System.getProperty("org.gradle.unsafe.executionhistory.readonly"), "true");
```

Comment:
This could, of course, be made static. My current set of opinions just likes not adding to the cost of loading classes, and this class is instantiated infrequently.